python/pyspark/sql/udaf.py

```diff
@@ -0,0 +1,897 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+"""
+User-defined aggregate function related classes and functions
+"""
+from typing import Any, TYPE_CHECKING, Optional, List, Iterator, Tuple
+
+from pyspark.sql.column import Column
+from pyspark.sql.types import (
+    DataType,
+    _parse_datatype_string,
+)
+from pyspark.errors import PySparkTypeError, PySparkNotImplementedError
+
+if TYPE_CHECKING:
+    from pyspark.sql._typing import DataTypeOrString, ColumnOrName
+    from pyspark.sql.dataframe import DataFrame
+
+__all__ = [
+    "Aggregator",
+    "UserDefinedAggregateFunction",
+    "udaf",
+]
+
+
+class Aggregator:
```

Comment:
do we necessarily need this class?
I see UDTF doesn't need a base class.

```
    >>> class TestUDTF:
    ...     def eval(self, *args: Any):
    ...         yield "hello", "world"
```

Reply:
we could do duck typing if not go with the inheritance. I think it is debatable and we could offer both solutions (with or without a base class).